File: Games/chieftain/src/data/fjord.ts
/**
 * 피오르드 해안 — v1의 유일한 맵 (GDD 6.2).
 *
 * **3×3, 아홉 칸.** 작게 만드는 것은 취향이 아니라 판을 끝내는 장치다
 * (GDD 4.1). 아홉 칸이면 땅이 5~6분이면 떨어지고, 그 뒤엔 서로를 미는 것
 * 말고 할 일이 없다.
 *
 * 칸은 물 위에 떠 있는 섬이고, 이웃한 칸 사이는 **다리 하나로만** 이어진다.
 * "피오르드는 지형이 좁은 통로를 만든다"는 GDD 6.2의 주장이 여기서 코드가
 * 된다 — 유닛은 칸 경계 아무 데로나 넘어가지 못하고 반드시 다리를 지난다.
 * 그래서 아홉 칸짜리 작은 맵에서도 "어느 목을 잡을 것인가"가 생긴다.
 *
 * ```
 *   [0,0]──[1,0]──[2,0]
 *     │      │      │
 *   [0,1]══[1,1]══[2,1]      ═ 는 본진에서 본진으로 가는 최단 경로
 *   본진     중앙    본진
 *     │      │      │
 *   [0,2]──[1,2]──[2,2]
 * ```
 *
 * 중앙 칸은 직행로의 유일한 길목이다. 돌아가려면 모서리 칸 두 개를 거쳐야
 * 하므로, 중앙을 잡은 쪽이 시간을 번다. 중앙에 중립을 두지 않은 것도 그
 * 때문이다 — 중앙은 **서로를 만나는 곳**이지 캠프를 도는 곳이 아니다.
 */
import { Rng } from '../core/rng'
import { TILE } from './tuning'
import type { NeutralKind } from './neutrals'

export const COLS = 3
export const ROWS = 3

/** 칸의 땅 부분 한 변. TILE보다 작아서 칸 사이에 물이 남는다. */
export const TILE_LAND = 28

export interface TileDef {
  readonly id: number
  readonly col: number
  readonly row: number
  /** 월드 좌표. 맵 중앙이 원점이다. */
  readonly x: number
  readonly z: number
}

export function tileId(col: number, row: number): number {
  return row * COLS + col
}

export function makeTiles(): TileDef[] {
  const tiles: TileDef[] = []
  for (let row = 0; row < ROWS; row++) {
    for (let col = 0; col < COLS; col++) {
      tiles.push({
        id: tileId(col, row),
        col,
        row,
        x: (col - (COLS - 1) / 2) * TILE,
        z: (row - (ROWS - 1) / 2) * TILE,
      })
    }
  }
  return tiles
}

/** 4방 인접. 대각선은 없다 — 다리가 네 변에만 놓이기 때문이다. */
export function neighborsOf(col: number, row: number): number[] {
  const out: number[] = []
  if (col > 0) out.push(tileId(col - 1, row))
  if (col < COLS - 1) out.push(tileId(col + 1, row))
  if (row > 0) out.push(tileId(col, row - 1))
  if (row < ROWS - 1) out.push(tileId(col, row + 1))
  return out
}

/** 두 칸을 잇는 다리의 위치 = 두 칸 중심의 중간점. */
export function bridgeBetween(a: TileDef, b: TileDef): { x: number; z: number } {
  return { x: (a.x + b.x) / 2, z: (a.z + b.z) / 2 }
}

/** 180° 회전 대칭의 짝 (GDD 4.2). */
export function mirrorOf(id: number): number {
  const col = id % COLS
  const row = Math.floor(id / COLS)
  return tileId(COLS - 1 - col, ROWS - 1 - row)
}

export const KEEP_P0 = tileId(0, 1)
export const KEEP_P1 = tileId(2, 1)
export const CENTER = tileId(1, 1)

/**
 * 중립 배치. **반쪽만 뽑고 회전 복사한다** (GDD 4.2).
 *
 * 왼쪽 절반의 빈 칸 세 개에 세 갈래를 하나씩 섞어 넣고, 그대로 오른쪽에
 * 회전시켜 놓는다. 그래서 매판 배치가 달라지지만 양쪽이 마주하는 기회는
 * 언제나 똑같다 — 승부는 운이 아니라 *누가 먼저 닿느냐*로 환원된다.
 *
 * 뽑는 칸이 셋이고 갈래도 셋이라 "무엇이 나오느냐"가 아니라 "어디에
 * 나오느냐"만 무작위다. 이건 의도한 것이다. 한 판에 세 갈래가 모두 등장해야
 * GDD 4.3이 말한 세 종류의 결정이 실제로 한 판 안에서 부딪힌다.
 */
export function placeNeutrals(seed: number): Map<number, NeutralKind> {
  const rng = new Rng(seed)

  // 왼쪽 절반에서 본진을 뺀 칸들. 오른쪽은 이들의 거울상이고, 중앙은 비운다.
  const half = [tileId(0, 0), tileId(0, 2), tileId(1, 0)]

  const kinds: NeutralKind[] = ['mercenary', 'creature', 'ruin']
  // Fisher–Yates. Rng를 쓰므로 같은 시드면 같은 배치가 나온다.
  for (let i = kinds.length - 1; i > 0; i--) {
    const j = rng.int(0, i)
    ;[kinds[i], kinds[j]] = [kinds[j]!, kinds[i]!]
  }

  const out = new Map<number, NeutralKind>()
  half.forEach((id, i) => {
    const kind = kinds[i]!
    out.set(id, kind)
    out.set(mirrorOf(id), kind)
  })
  return out
}

File: Games/chieftain-unity/Core/Data/Fjord.cs
using System;
using System.Collections.Generic;

namespace Chieftain.Core
{
    /// <summary>
    /// 피오르드 해안 — v1의 유일한 맵 (GDD 6.2). 3×3, 아홉 칸.
    ///
    /// <para>
    /// 칸은 물 위에 떠 있는 섬이고 이웃한 칸 사이는 다리 하나로만 이어진다.
    /// 그래서 아홉 칸짜리 작은 맵에서도 "어느 목을 잡을 것인가"가 생긴다.
    /// </para>
    /// </summary>
    public sealed class TileDef
    {
        public int Id;
        public int Col;
        public int Row;
        /// <summary>월드 좌표. 맵 중앙이 원점이다.</summary>
        public double X;
        public double Z;
    }

    public static class Fjord
    {
        public const int Cols = 3;
        public const int Rows = 3;

        /// <summary>칸의 땅 부분 한 변. TILE보다 작아서 칸 사이에 물이 남는다.</summary>
        public const double TileLand = 28;

        public static int TileId(int col, int row) => row * Cols + col;

        public static readonly int KeepP0 = TileId(0, 1);
        public static readonly int KeepP1 = TileId(2, 1);
        public static readonly int Center = TileId(1, 1);

        public static List<TileDef> MakeTiles()
        {
            var tiles = new List<TileDef>();
            for (int row = 0; row < Rows; row++)
            {
                for (int col = 0; col < Cols; col++)
                {
                    tiles.Add(new TileDef
                    {
                        Id = TileId(col, row),
                        Col = col,
                        Row = row,
                        X = (col - (Cols - 1) / 2.0) * Tuning.TILE,
                        Z = (row - (Rows - 1) / 2.0) * Tuning.TILE,
                    });
                }
            }
            return tiles;
        }

        /// <summary>4방 인접. 대각선은 없다 — 다리가 네 변에만 놓이기 때문이다.</summary>
        public static List<int> NeighborsOf(int col, int row)
        {
            var outp = new List<int>();
            if (col > 0) outp.Add(TileId(col - 1, row));
            if (col < Cols - 1) outp.Add(TileId(col + 1, row));
            if (row > 0) outp.Add(TileId(col, row - 1));
            if (row < Rows - 1) outp.Add(TileId(col, row + 1));
            return outp;
        }

        /// <summary>두 칸을 잇는 다리의 위치 = 두 칸 중심의 중간점.</summary>
        public static Vec2 BridgeBetween(TileDef a, TileDef b) =>
            new Vec2((a.X + b.X) / 2, (a.Z + b.Z) / 2);

        /// <summary>180° 회전 대칭의 짝 (GDD 4.2).</summary>
        public static int MirrorOf(int id)
        {
            int col = id % Cols;
            int row = id / Cols;
            return TileId(Cols - 1 - col, Rows - 1 - row);
        }

        /// <summary>
        /// 중립 배치. **반쪽만 뽑고 회전 복사한다** (GDD 4.2).
        ///
        /// <para>
        /// 뽑는 칸이 셋이고 갈래도 셋이라 "무엇이 나오느냐"가 아니라 "어디에
        /// 나오느냐"만 무작위다. 한 판에 세 갈래가 모두 등장해야 GDD 4.3이 말한
        /// 세 종류의 결정이 실제로 한 판 안에서 부딪힌다.
        /// </para>
        /// </summary>
        public static Dictionary<int, NeutralKind> PlaceNeutrals(uint seed)
        {
            var rng = new Rng(seed);

            // 왼쪽 절반에서 본진을 뺀 칸들. 오른쪽은 이들의 거울상이고, 중앙은 비운다.
            int[] half = { TileId(0, 0), TileId(0, 2), TileId(1, 0) };

            var kinds = new[] { NeutralKind.Mercenary, NeutralKind.Creature, NeutralKind.Ruin };
            // Fisher–Yates. Rng를 쓰므로 같은 시드면 같은 배치가 나온다.
            for (int i = kinds.Length - 1; i > 0; i--)
            {
                int j = rng.Int(0, i);
                (kinds[i], kinds[j]) = (kinds[j], kinds[i]);
            }

            var outp = new Dictionary<int, NeutralKind>();
            for (int i = 0; i < half.Length; i++)
            {
                var kind = kinds[i];
                outp[half[i]] = kind;
                outp[MirrorOf(half[i])] = kind;
            }
            return outp;
        }
    }
}
